File: BT/BipartiteGraph/BipartiteGraph.java
import java.util.*;

/*
Black bucket
Red bucket


*/
class Solution {
    public boolean solve(int[][] graph) {
        if (graph.length == 0 || graph[0].length == 0) return false;

        Set<Integer> bucket1 = new HashSet<>();
        Set<Integer> bucket2 = new HashSet<>();

        for (int node = 0; node < graph.length; node++) {
            if (isVisited(node, bucket1, bucket2)) continue;
            if (!solve(graph, node, bucket1, bucket2, true)) return false;
        }
        return true;
    }

    private boolean solve(int[][] graph, int node, Set<Integer> bucket1, Set<Integer> bucket2, boolean isBucket1) {
        if (!isValid(node, bucket1, bucket2, isBucket1)) return false;

        if (isBucket1) bucket1.add(node);
        else bucket2.add(node);

        for (int neighbor : graph[node]) {
            if (!isValid(neighbor, bucket1, bucket2, !isBucket1)) return false;
            if (isVisited(neighbor, bucket1, bucket2)) continue;
            if (!solve(graph, neighbor, bucket1, bucket2, !isBucket1)) return false;
        }

        return true;
    }

    private boolean isVisited(int node, Set<Integer> bucket1, Set<Integer> bucket2) {
        return bucket1.contains(node) || bucket2.contains(node);
    }

    private boolean isValid(int node, Set<Integer> bucket1, Set<Integer> bucket2, boolean isBucket1) {
        if (!isVisited(node, bucket1, bucket2)) return true;
        return (isBucket1 && bucket1.contains(node) || !isBucket1 && bucket2.contains(node));
    }
}

File: BT/BipartiteGraph/BipartiteGraph.py
class Solution:
    def solve(self, graph):
        if not graph:
            return True

        reds, blacks = set(), set()
        for start_node in range(len(graph)):
            if start_node in reds or start_node in blacks:
                continue
            if not self.dfs(graph, start_node, True, blacks, reds):
                return False
        return True

    def dfs(self, graph, node, is_black, blacks, reds):
        if node in blacks or node in reds:
            return (is_black and  node in blacks) or (not is_black and node in reds)
        
        if is_black:
            blacks.add(node)
        else:
            reds.add(node)
        
        for child in graph[node]:
            if not self.dfs(graph, child, not is_black, blacks, reds):
                return False
        
        return True
